### DISCUSSION TITLE
Build AppImage inside of Flatpak Runtime

### DISCUSSION THREAD

**Discussion by pktiuk** [AppImage/AppImageKit#1292] [Q&A]
Hello.   
Has anyone tried building AppImage packages inside flatpak runtimes?   
I am trying to do it with an app, which was developed using flatpak runtime and due to its fresh dependencies building it using libs from repositories is not doable (and I don't  want to compile everything manually) [link to related issue](https://github.com/nokyan/resources/issues/20).    
I think using a Flatpak runtime, just like we could use Docker may be a good idea.   
Has anyone tried doing it earlier (I have some technical problems with this)? What do you think about this idea?

**Comment by probonopd** [comment]
Hello @pktiuk, what would be the advantage of building AppImages inside Flatpak runtimes?

Would you want to ship parts of the Flatpak runtime inside the AppImages?

**Reply by pktiuk** [reply]
I just want to package an app with compatible libraries (which are part of this runtime).    
The main advantage is fact, that all the dependencies in correct versions are already installed inside of this runtime.
And exactly the same libraries will be used in another packaging format (flatpak).

**Reply by probonopd** [reply]
Well, the key point is that what is inside the AppImage needs no more recent versions of the libraries than what is on the outside (user's target systems), unless you package _all_ required libraries inside the AppImage.

So, I don't see why using the Flatpak runtime for building the ingredients of an AppImage would be beneficial (especially since we cannot assume the Flatpak runtime to be present on all target systems, and even much less in what version). But then, I don't know much about Flatpak to begin with.

**Reply by pktiuk** [reply]
I just want to use flatpak, as you could use a docker in this case.    
You build an app inside, and then create AppImage file (containing your app and required libraries). Then you can copy this file and launch it outside flatpak runtime, because it already contains all the required dependencies which are not present on the host system.

**Reply by probonopd** [reply]
So you would want to bundle _all_ required libraries which the applications (down to, and including glibc) from the Flatpak runtime? That would probably work (disclaimer: I have never tried it). But then, you could do that with any other host build system which basically the same result.

In case you don't want to bundle _all_ required libraries, you'd have to ensure to use a Flatpak runtime that is older than the oldest still-supported target system (Linux distribution release).

**Comment by TheAssassin** [comment]
I don't see a reason why you couldn't do it. If done right (i.e., build on old, run on new systems), your resulting AppImage will work just as expected. I highly encourage you to try it out yourself. Please report bugs if you find any.
